torch/csrc/jit/codegen/cuda/runtime/block_welford_outer.cu

```diff
@@ -0,0 +1,265 @@
+      }
+      auto peer_avg = __shfl_xor_sync(0xffffffff, pushed_avg, lane_mask);
+      auto peer_var = __shfl_xor_sync(0xffffffff, pushed_var, lane_mask);
+
+      auto delta = peer_avg - self_avg;
+      self_avg += delta * b_N_div_ab_N;
+      self_var += peer_var + delta * delta * ((DataType)(inp_N)) * b_N_div_ab_N;
+
+      inp_avg[index_in_chunk] = self_avg;
+      inp_var[index_in_chunk] = self_var;
+    }
+    inp_N = updated_N;
+  }
+
+  // At this point, chunk_size is reduced to NumVals/wdimy as
+  // mentioned above. Each thread has warp-reduced chunk_size values
+  // in array inp. This chunk_size_post_reduction should be equal to
+  // chunk_size at this point.
+  constexpr int chunk_size_post_reduction = NumVals / wdimy;
+
+  // More specifically, the warp_tidy of each thread defines
+  // the chunk IDs held by the thread as follows:
+  //
+  // [warp_tidy * chunk_size_post_reduction, warp_tidy *
+  // chunk_size_post_reduction + chunk_size_post_reduction]
+  //
+  // Each thread uploads the chunk_size_post_reduction values one by
+  // one. Each chunk is spread by BDIMX * BDIMY values. The data
+  // layout of the shared memory is:
+  //
+  // [chunk_size, wid, warp_tidy, TIDx]
+  //
+  // The remaining reduction is done on the WID
+  // dimension. More specifically, we assign one warp per chunk (or
+  // a value of the group). The wdimy threads of the same threadId.x
+  // collectively reduce num_warps partial results, each of which is
+  // stored with stride 32. This means that there will be wdimy-way
+  // bank conflicts, so to avoid that, swizziling is also employed.
+#pragma unroll
+  for (int i = 0; i < chunk_size; ++i) {
+    // Accumulating smem offset from the innermost dimension
+    int smem_offset = 0;
+    // TIDx
+    smem_offset += threadIdx.x;
+    // Warp_TIDy with swizzle
+    smem_offset += ((warp_tidy + wid) % wdimy) * BDIMX;
+    // WID
+    smem_offset += wid * 32;
+    // chunk_size
+    smem_offset += i * BDIMX * BDIMY;
+    smem[smem_offset] = inp_avg[i];
+    smem[smem_var_offset + smem_offset] = inp_var[i];
+    // Upload N only when threadIdx.x == 0 && chunk_index == 0
+    if (threadIdx.x == 0 && i == 0 && warp_tidy == 0) {
+      reinterpret_cast<nvfuser_index_t*>(smem + smem_N_offset)[wid] = inp_N;
+    }
+  }
+
+  __syncthreads();
+
+  // The next step is to let each thread of a warp independently
+  // accumulate the partial results on the shared memory
+  // reduction. A single warp is used to accumulate of the partial
+  // results for a single chunk, so warp wid takes care of the wid-th
+  // chunk.
+  //
+  // The starting offset of partial results of a chunk is:
+  //
+  // (wid % chunk_size_post_reduction) * BDIMX * BDIMY + (wid /
+  // chunk_size_post_reduction) * BDIMX
+  //
+  // Note that each thread had chunk_size_post_reduction contiguous
+  // chunks, so when uploaded to shmem, they are strided by
+  // BDIMX*BDIMY, hence (wid % chunk_size_post_reduction) * BDIMX *
+  // BDIMY.
```

Comment:
This I'm missing how you got `wid % chunk_size`. I guess you're just wrapping wid instead of truncating threads off, is this for syncthreads usage?

Reply:
It's because each warp stored `chunk_size_post_reduction` group entries to the smem buffer, and each chunk is separated by `BDIMX * BDIMY` values. For warps of `wid < chunk_size_post_reduction`, each warp reads the values stored by threads of `warp_tidy == 0` and  the starting position is at `wid * BDIMX * BDIMY`. The next set of `chunk_size_post_reduction` warps reads the values stored by threads of `warp_tidy == 1`, so  in addition to `(wid % chunk_size_post_reduction) * BDIMX * BDIMY`, an additional offset of `(wid / chunk_size_post_reduction) * BDIMX` is used.

Reply:
Another thing that I should have written explicitly would be that, given `[chunk_size, wid, warp_tidy, TIDx]`, its group offset is `chunk_size_index + warp_tidy * chunk_size`.

In our notation, this should look like:
(TIDx is replaced with warp_tidx as it's also the X-dim index within each warp)
```
[chunk_size, wid, warp_tidy, warp_tidx]
split wid by wdimy -> [chunk_size, wid / wdimy, wdimy, warp_tidy, warp_tidx]
merge wdimy and warp_tidx -> [chunk_size, wid / wdimy, warp_tidy, wdimy * warp_tidx]
merge warp_tidy and chunk_size -> [wid / wdimy, warp_tidy * chunk_size, wdimy * warp_tidx]
parallelize with WID (warp ID) -> [wid / wdimy, warpId(warp_tidy * chunk_size), wdimy * warp_tidx]
parallelize with LID (lane ID) -> [wid / wdimy, warpId(warp_tidy * chunk_size), laneId(wdimy * warp_tidx)]
```